format/src/muxer.rs

```diff
@@ -2,40 +2,59 @@ use crate::common::*;
 use crate::data::packet::Packet;
 use crate::data::value::*;
 use std::any::Any;
-use std::io::{ErrorKind, Seek, SeekFrom, Write};
+use std::io::{Cursor, ErrorKind, Seek, SeekFrom, Write};
 use std::sync::Arc;
 
 use crate::error::*;
 
-pub trait WriteSeek: Write + Seek {}
+/// A trait for a non-seekable object.
+pub trait WriteOwned: Write + ToOwned {}
+/// A trait for a seekable object.
+pub trait WriteSeek: Write + Seek + ToOwned {}
 
-impl<T: Write + Seek> WriteSeek for T {}
+impl<T: Write + ToOwned> WriteOwned for T {}
+impl<T: Write + Seek + ToOwned> WriteSeek for T {}
 
 /// Runtime wrapper around either a [`Write`] or a [`WriteSeek`] trait object which supports querying
 /// for seek support.
-pub enum Writer {
-    NonSeekable(Box<dyn Write>, u64),
-    Seekable(Box<dyn WriteSeek>),
+pub enum Writer<WO = Cursor<Vec<u8>>, WS = Cursor<Vec<u8>>> {
+    NonSeekable(WO, u64),
+    Seekable(WS),
 }
 
-impl Writer {
-    /// Creates a [`Writer`] from a trait object that implements both [`Write`] and [`Seek`].
-    pub fn from_seekable(inner: Box<dyn WriteSeek>) -> Self {
-        Self::Seekable(inner)
+impl<WO: WriteOwned> Writer<WO, Cursor<Vec<u8>>> {
+    /// Creates a [`Writer`] from an object that implements the [`Write`] trait.
+    pub fn from_nonseekable(inner: WO) -> Self {
+        Self::NonSeekable(inner, 0)
```

Comment:
You may implement a blanket From/Into in the WriteSeek/WriteOwned and call it from a specialized From for `Writer<WO, Cursor<>>/Writer<Cursor<>,WS` maybe?